File: Python/Algorithms/654.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def constructMaximumBinaryTree(self, nums):
        # use scopes to access nums.
        def build_tree(start, end):
            # finds max value and its index.
            max_value, idx = -1, start
            for i in range(start, end):
                if nums[i] > max_value:
                    max_value, idx = nums[i], i

            # go left/right if elements will exist
            # in those ranges.
            left = build_tree(start, idx) if idx > start else None
            idx = idx + 1
            right = build_tree(idx, end) if end > idx else None

            # add left, right and return Node.
            return TreeNode(max_value, left, right)
        return build_tree(0, len(nums))

File: Rust/Algorithms/654.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    pub fn construct_maximum_binary_tree(
        nums: Vec<i32>,
    ) -> Option<Rc<RefCell<TreeNode>>> {
        fn build_tree(
            nums: &Vec<i32>,
            start: usize,
            end: usize,
        ) -> Option<Rc<RefCell<TreeNode>>> {
            let (mut val, mut id) = (nums[start], start);
            for i in start + 1..end {
                if nums[i] > val {
                    val = nums[i];
                    id = i;
                }
            }

            let (mut left, mut right) = (None, None);
            if id > start {
                left = build_tree(nums, start, id);
            }
            if end > id + 1 {
                right = build_tree(nums, id + 1, end);
            }
            Some(Rc::new(RefCell::new(TreeNode { val, left, right })))
        }
        build_tree(&nums, 0, nums.len())
    }
}
